python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -63,3 +66,205 @@ def get_instance_manager_state(
             The reply to the request.
         """
         pass
+
+
+class InstanceUtil:
+    """
+    A helper class to group updates and operations on an Instance object defined
+    in instance_manager.proto
+    """
+
+    @staticmethod
+    def new_instance(
+        instance_id: str,
+        instance_type: str,
+        resources: Dict[str, float],
+        request_id: str = "",
+    ) -> Instance:
+        instance = Instance()
+        instance.version = 0  # it will be populated by the underlying storage.
+        instance.instance_id = instance_id
+        instance.instance_type = instance_type
+        instance.launch_request_id = request_id
+        instance.total_resources.update(resources)
+        instance.status = Instance.QUEUED
+        InstanceUtil._record_status_transition(
+            instance, Instance.QUEUED, "created from InstanceUtil"
+        )
+        return instance
+
+    @staticmethod
+    def is_cloud_instance_allocated(instance: Instance) -> bool:
+        """
+        Returns True if the instance is in a status where there could exist
+        a cloud instance allocated by the cloud provider.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in {
+            Instance.ALLOCATED,
+            Instance.RAY_INSTALLING,
+            Instance.RAY_RUNNING,
+            Instance.RAY_STOPPING,
+            Instance.RAY_STOPPED,
+            Instance.STOPPING,
+            Instance.RAY_INSTALL_FAILED,
+        }
+
+    @staticmethod
+    def is_ray_running_reachable(instance: Instance) -> bool:
+        """
+        Returns True if the instance is in a status where it may transition
+        to RAY_RUNNING status.
+        """
+        assert instance.status != Instance.UNKNOWN
+        return instance.status in [
```

Comment:
should this be dynamically calculated based on get_valid_transitions? We can calculate once and cache the result

Reply:
(or if you plan to keep this guarantee via unit test, that sounds good to me too)

Reply:
it's kept in unit testing now.